src/jit/optimizer.cpp

```diff
@@ -1790,7 +1774,268 @@ void Compiler::optFindNaturalLoops()
                 }
 #endif // FEATURE_EH_FUNCLETS && defined(_TARGET_ARM_)
 
-                /* At this point we have a loop - record it in the loop table
+                // Compact the loop (sweep through it and move out any blocks that aren't part of the
+                // flow cycle), and count the exits.
+                BasicBlock* moveAfter = nullptr;
+                for (BasicBlock* previous = top->bbPrev; previous != bottom;)
+                {
+                    BasicBlock* block = previous->bbNext;
```

Comment:
I wonder if this would be a clearer if it was conceptually split up as follows (with appropriate bits of state passed in and out):
```
for block in range [top..bottom]
   previous = block->bbPrev;
   if block is in loop: CountLoopExit(...); next = block->bbNext
   if block is not in loop: next = MoveBlockRangeOutOfLoopOrTolerateInLoop(...)  // ret null if fails
```